src/aggregators/direct.jl

```diff
@@ -1,68 +1,94 @@
-type DirectJumpAggregation{T,F1,F2} <: AbstractJumpAggregator
-  next_jump::T
+type DirectJumpAggregation{T,F1,F2,RNG} <: AbstractSSAJumpAggregator
+  next_jump::Int
+  next_jump_time::T
   end_time::T
   cur_rates::Vector{T}
   sum_rate::T
   rates::F1
   affects!::F2
   save_positions::Tuple{Bool,Bool}
+  rng::RNG
 end
 
-@inline function (p::DirectJumpAggregation)(u,t,integrator) # condition
-  p.next_jump==t
+########### The following routines should be templates for all SSAs ###########
+
+# condition for jump to occur
+@inline function (p::DirectJumpAggregation)(u, t, integrator) 
+  p.next_jump_time == t
 end
 
-function (p::DirectJumpAggregation)(integrator) # affect!
-  rng_val = rand()
-  i = searchsortedfirst(p.cur_rates,rng_val)
-  @inbounds p.affects![i](integrator)
-  p.sum_rate,ttnj = time_to_next_jump(integrator.u,integrator.p,integrator.t,p.rates,p.cur_rates)
-  p.next_jump = integrator.t + ttnj
-  if p.next_jump < p.end_time
-    add_tstop!(integrator,p.next_jump)
-  end
+# executing jump at the next jump time
+function (p::DirectJumpAggregation)(integrator) 
+  execute_jumps!(p, integrator, integrator.u, integrator.p, integrator.t)
+  generate_jumps!(p, integrator, integrator.u, integrator.p, integrator.t)
+  register_next_jump_time!(integrator, p, integrator.t)
   nothing
 end
 
-function (p::DirectJumpAggregation)(dj,u,t,integrator) # initialize
-  sum_rate,next_jump = time_to_next_jump(u,integrator.p,t,p.rates,p.cur_rates)
-  p.sum_rate = sum_rate
-  p.next_jump = t + next_jump
-  if p.next_jump < p.end_time
-    add_tstop!(integrator,p.next_jump)
-  end
+# setting up a new simulation
+function (p::DirectJumpAggregation)(dj, u, t, integrator) # initialize
+  initialize!(p, integrator, u, integrator.p, t)
+  register_next_jump_time!(integrator, p, t)
   nothing
 end
 
-function time_to_next_jump(u,p,t,rates,cur_rates)
-  @inbounds fill_cur_rates(u,p,t,cur_rates,1,rates...)
-  sum_rate = sum(cur_rates)
-  @fastmath normalizer = 1/sum_rate
-  @inbounds cur_rates[1] = normalizer*cur_rates[1]
-  @inbounds for i in 2:length(cur_rates) # normalize for choice, cumsum
-    cur_rates[i] = normalizer*cur_rates[i] + cur_rates[i-1]
-  end
-  sum_rate,randexp()/sum_rate
+############################# Required Functions #############################
+
+# creating the JumpAggregation structure
+@inline function aggregate(aggregator::Direct, u, p, t, end_time, constant_jumps, save_positions, rng)
```

Comment:
The inline here also seems unnecessary (it was present before but I don't see what good it can do here).

Reply:
Yeah it's probably unnecessary.